### PR TITLE
[QEMU] Add number of cores, target list for build

### PR DESCRIPTION
**Pull request by mehrdadh**
This PR:
- Adds number of core variable to docker/install/ubuntu_install_qemu script
- And adds target list as optional input argument

### CODE REVIEW

**Comment by mehrdadh** [comment]
cc @areusch @leandron

File: docker/install/ubuntu_install_qemu.sh [hunk 1]
```diff
@@ -45,7 +51,7 @@ gpg --verify qemu-5.1.0.tar.xz.sig
 tar -xf qemu-5.1.0.tar.xz
 cd qemu-5.1.0
 ./configure --target-list=aarch64-softmmu,arm-softmmu,i386-softmmu,riscv32-softmmu,riscv64-softmmu,x86_64-softmmu
-make -j2
+make -j${NUM_CORE}
```
Comment (tkonolige) [inline docker/install/ubuntu_install_qemu.sh:54]: Why not use `nproc`?

Comment (mehrdadh) [reply docker/install/ubuntu_install_qemu.sh:54]: Because some people are building this on a local machine to test with low number of cores and it could crash.

File: docker/install/ubuntu_install_qemu.sh [hunk 2]
```diff
@@ -16,22 +16,46 @@
...
+#   --target-list is list of target for QEMU comma seperated. e.g. aarch64-softmmu,arm-softmmu,...
+#
+
+set -e
+set -o pipefail
+
+QEMU_DIR=qemu-5.1.0
+
+# Get number of cores for build
 if [[ "${CI_NUM_CORE}" ]]; then
```
Comment (areusch) [inline docker/install/ubuntu_install_qemu.sh:32]: let's pick either CORE or CORES and stick with it. Prefer NUM_CORES since it's more natural

File: docker/install/ubuntu_install_qemu.sh [hunk 3]
```diff
@@ -16,22 +16,46 @@
...
 
-set -e
-set -u
-set -o pipefail
+# Set target list for QEMU
+if [ "$1" == "--target-list" ]; then
+    shift
+    target_list=$1
+else
+    target_list="aarch64-softmmu,arm-softmmu,i386-softmmu,riscv32-softmmu,riscv64-softmmu,x86_64-softmmu"
```
Comment (areusch) [inline docker/install/ubuntu_install_qemu.sh:43]: comment where this list comes from

File: docker/install/ubuntu_install_qemu.sh [hunk 4]
```diff
@@ -16,22 +16,46 @@
...
+fi
 
-sudo sed -i '/deb-src/s/^# //' /etc/apt/sources.list
-apt update
-apt-get -y build-dep qemu
+# Check if QEMU already built
+rebuild=0
```
Comment (areusch) [inline docker/install/ubuntu_install_qemu.sh:47]: why do we need this logic in this file?

File: docker/install/ubuntu_install_qemu.sh [hunk 5]
```diff
@@ -16,22 +16,46 @@
...
+if [ -d ${QEMU_DIR} ]; then
+    rebuild=1
+fi
+
+if [ $rebuild -eq 0 ]; then
+    sudo sed -i '/deb-src/s/^# //' /etc/apt/sources.list
```
Comment (areusch) [inline docker/install/ubuntu_install_qemu.sh:53]: i think you should at least filter by the particular deb-src line you want to re-enable? also comment why this is needed

File: docker/install/ubuntu_install_qemu.sh [hunk 6]
```diff
@@ -45,14 +69,16 @@ p5ez/+2k4VAIwIQoP5DoO06waLBffvLIAdPPKYsx71K67OoGG2svc7duC/+5qf1x
...
-tar -xf qemu-5.1.0.tar.xz
 cd qemu-5.1.0
-./configure --target-list=aarch64-softmmu,arm-softmmu,i386-softmmu,riscv32-softmmu,riscv64-softmmu,x86_64-softmmu
-make -j${NUM_CORE}
+./configure --target-list=${target_list}
+make -j${num_cores}
 sudo make install
 
-# For debugging with qemu
+# # For debugging with qemu
```
Comment (areusch) [inline docker/install/ubuntu_install_qemu.sh:83]: nit: revert this change

File: docker/install/ubuntu_install_qemu.sh [hunk 7]
```diff
@@ -45,14 +69,16 @@ p5ez/+2k4VAIwIQoP5DoO06waLBffvLIAdPPKYsx71K67OoGG2svc7duC/+5qf1x
...
+    gpg --verify qemu-5.1.0.tar.xz.sig
+
+    tar -xf qemu-5.1.0.tar.xz
+fi
 
-tar -xf qemu-5.1.0.tar.xz
 cd qemu-5.1.0
-./configure --target-list=aarch64-softmmu,arm-softmmu,i386-softmmu,riscv32-softmmu,riscv64-softmmu,x86_64-softmmu
-make -j${NUM_CORE}
+./configure --target-list=${target_list}
```
Comment (areusch) [inline docker/install/ubuntu_install_qemu.sh:79]: seems like if rebuild should also gate these? should it be renamed?

**Review by areusch — CHANGES_REQUESTED** [review]

File: docker/install/ubuntu_install_qemu.sh (hunk 2 above)
Comment (mehrdadh) [reply docker/install/ubuntu_install_qemu.sh:32]: done.

File: docker/install/ubuntu_install_qemu.sh (hunk 4 above)
Comment (mehrdadh) [reply docker/install/ubuntu_install_qemu.sh:47]: The reason is that we could update the RVM on vagrant cloud with a version that has qemu already installed with limited targets. And then if someone is using this locally, it will only rebuild with new set of targets.

File: docker/install/ubuntu_install_qemu.sh (hunk 7 above)
Comment (mehrdadh) [reply docker/install/ubuntu_install_qemu.sh:79]: no, because the target list could be updated.

File: docker/install/ubuntu_install_qemu.sh (hunk 4 above)
Comment (areusch) [reply docker/install/ubuntu_install_qemu.sh:48]: ah i see. okay, do you intend to push that update soon? otherwise, may be better to keep this file simple. i agree it's a nice feature, but wondering how much extra it adds to the base image.

File: docker/install/ubuntu_install_qemu.sh (hunk 7 above)
Comment (mehrdadh) [reply docker/install/ubuntu_install_qemu.sh:69]: removed the rebuild flag.

File: docker/install/ubuntu_install_qemu.sh [hunk 8]
```diff
@@ -16,10 +16,32 @@
...
+# Usage: ubuntu_install_qemu.sh [--target-list target0,target1,...]
+#   --target-list is list of target for QEMU comma seperated. e.g. aarch64-softmmu,arm-softmmu,...
+#
+
 set -e
-set -u
 set -o pipefail
 
+# Get number of cores for build
+if [[ "${TVM_CI_NUM_CORES}" ]]; then
```
Comment (areusch) [inline docker/install/ubuntu_install_qemu.sh:30]: i think you need e.g. `if [ -n "${TVM_CI_NUM_CORES}" ]` or some operator, no?

**Review by areusch — COMMENTED** [review]
ok sorry one more question, then looks good!

File: docker/install/ubuntu_install_qemu.sh (hunk 8 above)
Comment (mehrdadh) [reply docker/install/ubuntu_install_qemu.sh]: shoot, it was working without `-n`, that's why I missed it. But this is a better syntax.
Thanks for finding this!

**Review by areusch — APPROVED** [review]

**Comment by areusch** [comment]
thanks @mehrdadh , the PR is now merged!
